File: app/src/main/java/com/maple/recordwav/ui/MainActivity.java
package com.maple.recordwav.ui;

import android.Manifest;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TabHost;
import android.widget.TextView;

import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentTabHost;

import com.maple.recordwav.R;
import com.maple.recordwav.utils.T;
import com.maple.recordwav.utils.permission.PermissionFragment;
import com.maple.recordwav.utils.permission.PermissionListener;

/**
 * @author maple
 * @time 2018/4/8.
 */
public class MainActivity extends FragmentActivity {
    TextView mTitle;
    FragmentTabHost mTabHost;

    // Fragment界面
    private Class[] fragmentArray = {RecordPage.class, PlayPage.class, ParsePage.class};
    // 选项卡图片
    private int[] mImageViewArray = {R.drawable.tab_record_icon, R.drawable.tab_play_icon, R.drawable.tab_parse_icon};
    // 选项卡文字
    String[] mTextViewArray;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mTitle = findViewById(R.id.tv_title);
        mTabHost = findViewById(R.id.tabhost);

        mTextViewArray = getResources().getStringArray(R.array.tab_fun_array);

        initView();

        PermissionFragment.getPermissionFragment(this)
                .setPermissionListener(new PermissionListener() {
                    @Override
                    public void onPermissionGranted() {

                    }

                    @Override
                    public void onPermissionDenied(String[] deniedPermissions) {
                        T.showShort(MainActivity.this, "不同意将无法使用");
                    }
                })
                .checkPermissions(new String[]{
                        Manifest.permission.READ_EXTERNAL_STORAGE,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        Manifest.permission.RECORD_AUDIO
                });
    }

    private void initView() {
        mTabHost.setup(this, getSupportFragmentManager(), R.id.fl_content);
        for (int i = 0; i < fragmentArray.length; i++) {
            TabHost.TabSpec tabSpec = mTabHost.newTabSpec(mTextViewArray[i]).setIndicator(getTabItemView(i));
            mTabHost.addTab(tabSpec, fragmentArray[i], null);
            mTabHost.getTabWidget().getChildAt(i).setBackgroundResource(R.drawable.selector_tab_background);
        }

        mTabHost.setOnTabChangedListener(new TabHost.OnTabChangeListener() {
            @Override
            public void onTabChanged(String tag) {
                mTitle.setText(tag);
            }
        });
        mTabHost.setCurrentTab(0);
        mTitle.setText(mTextViewArray[0]);
    }


    private View getTabItemView(int index) {
        View view = LayoutInflater.from(this).inflate(R.layout.tab_item_view, null);

        ImageView imageView = view.findViewById(R.id.imageview);
        imageView.setImageResource(mImageViewArray[index]);

        TextView textView = view.findViewById(R.id.textview);
        textView.setText(mTextViewArray[index]);

        return view;
    }
}

File: app/src/main/java/com/maple/recordwav/ui/MainActivity.kt
package com.maple.recordwav.ui

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.FragmentActivity
import com.google.android.material.tabs.TabLayout
import com.maple.msdialog.AlertDialog
import com.maple.recordwav.R
import com.maple.recordwav.databinding.ActivityMainBinding
import com.maple.recordwav.utils.BottomTabView
import com.maple.recordwav.utils.FragmentChangeManager
import com.maple.recordwav.utils.permission.RxPermissions
import com.maple.recordwav.utils.setContentViewAndSetWindowInsets


/**
 * Record Wav 示例（Kotlin版）
 *
 * @author maple
 * @time 2018/4/8.
 */
class MainActivity : FragmentActivity() {
    private lateinit var binding: ActivityMainBinding
    private var tabIndex: Int = 0 // 当前显示页面

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentViewAndSetWindowInsets(binding.root, binding.flTopBar)

        initTitleBar()
        initView()
        requestPermission()
    }

    private fun initTitleBar() {
        binding.tvTitle.text = "Wav 助手(Kotlin)"
        binding.tvSwitch.text = "转Java\n版本"
        binding.tvSwitch.setOnClickListener {
            val intent = Intent(this@MainActivity, MainActivityJava::class.java)
            startActivity(intent)
            finish()
        }
    }

    private fun initView() {
        val fragmentList = arrayListOf(
            supportFragmentManager.findFragmentByTag(FragmentChangeManager.getFragmentTag(0)) ?: RecordPage(),
            supportFragmentManager.findFragmentByTag(FragmentChangeManager.getFragmentTag(1)) ?: PlayParsePage(),
            supportFragmentManager.findFragmentByTag(FragmentChangeManager.getFragmentTag(2)) ?: AboutPage()
        )
        val iconArr = arrayListOf(R.drawable.sel_tab_record_icon, R.drawable.sel_tab_play_icon, R.drawable.sel_tab_parse_icon)
        val titleArr = arrayListOf(getString(R.string.record), getString(R.string.play), getString(R.string.info))
        fragmentList.forEachIndexed { index, _ ->
            val tab: TabLayout.Tab = binding.tlTab.newTab()
            tab.customView = BottomTabView(this).apply {
                setIcon(iconArr[index], null)
                setTitle(titleArr[index])
            }
            binding.tlTab.addTab(tab)
        }
        val fgManager = FragmentChangeManager(supportFragmentManager, R.id.fl_content, fragmentList, tabIndex)
        binding.tlTab.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                tabIndex = tab.position
                fgManager.setCurrentFragment(tabIndex)
                (tab.customView as BottomTabView?)?.setSelectStatus(true)
            }

            override fun onTabReselected(tab: TabLayout.Tab?) {}
            override fun onTabUnselected(tab: TabLayout.Tab?) {
                (tab?.customView as BottomTabView?)?.setSelectStatus(false)
            }
        })
        binding.tlTab.getTabAt(tabIndex)?.select()
    }

    @SuppressLint("CheckResult")
    private fun requestPermission() {
        RxPermissions(this).request(
            Manifest.permission.RECORD_AUDIO
        ).subscribe { granted ->
            if (!granted) {
                AlertDialog(this).apply {
                    setCancelable(false)
                    setCanceledOnTouchOutside(false)
                    setTitle("权限不足！")
                    setMessage("录音必须要有“RECORD_AUDIO”权限哦，否则无法录音和存储。")
                    setLeftButton("退出") { this@MainActivity.finish() }
                    setRightButton("再选一次") { requestPermission() }
                }.show()
            }
        }
    }
}